x/apps/keeper/appStateChanges_test.go

```diff
@@ -371,3 +374,137 @@ func TestAppStateChange_BeginUnstakingApplication(t *testing.T) {
 		})
 	}
 }
+
+// Fund an account by minting tokens in a module account and sending it
+// to the given account.
+func fundAccount(
+	t *testing.T,
+	ctx *sdk.Context,
+	k *Keeper,
+	address sdk.Address,
+	amount sdk.BigInt,
+) {
+	minter := types.StakedPoolName
+	addMintedCoinsToModule(t, ctx, k, minter)
+	sendFromModuleToAccount(t, ctx, k, minter, address, amount)
+}
+
+func transferApp(
+	t *testing.T,
+	ctx *sdk.Context,
+	k *Keeper,
+	transferFrom, transferTo crypto.PublicKey,
+) sdk.Error {
+	curApp, err := k.ValidateApplicationTransfer(
+		ctx,
+		transferFrom,
+		types.MsgStake{
+			PubKey: transferTo,
+			Chains: nil,
+			Value:  sdk.ZeroInt(),
+		},
+	)
+	if err != nil {
+		return err
+	}
+	return k.TransferApplication(ctx, curApp, transferTo)
+}
+
+func TestAppStateChange_Transfer(t *testing.T) {
+	originalUpgradeHeight := codec.UpgradeHeight
+	originalTestMode := codec.TestMode
+	originalFeatKey := codec.UpgradeFeatureMap[codec.AppTransferKey]
+	t.Cleanup(func() {
+		codec.UpgradeHeight = originalUpgradeHeight
+		codec.TestMode = originalTestMode
+		codec.UpgradeFeatureMap[codec.AppTransferKey] = originalFeatKey
+	})
+	codec.UpgradeHeight = -1
+
+	ctx, _, keeper := createTestInput(t, true)
+
+	apps := make([]types.Application, 3)
+	pubKeys := make([]crypto.PublicKey, 4)
+	addrs := make([]sdk.Address, 4)
+	for i := range apps {
+		apps[i] = createNewApplication()
+		pubKeys[i] = apps[i].PublicKey
+		addrs[i] = sdk.Address(pubKeys[i].Address())
+		amount := sdk.NewInt(int64(i) * 10000000)
+		fundAccount(t, &ctx, &keeper, apps[i].Address, amount)
+		assert.Nil(t, keeper.StakeApplication(ctx, apps[i], amount))
+	}
+	pubKeys[3] = getRandomPubKey()
+	addrs[3] = sdk.Address(pubKeys[3].Address())
+
+	// apps[0]: staked
+	// apps[1]: unstaking
+	// apps[2]: staked and jailed
+	keeper.BeginUnstakingApplication(ctx, apps[1])
+	keeper.JailApplication(ctx, apps[2].Address)
+
+	stakedApps := keeper.GetApplications(ctx, uint16(len(apps)*2))
+	assert.Equal(t, len(apps), len(stakedApps))
+
+	// transfer: apps[0]-->apps[3](new) fails before ugprade
+	err := transferApp(t, &ctx, &keeper, pubKeys[0], pubKeys[3])
+	assert.NotNil(t, err)
+	assert.Equal(t, keeper.Codespace(), err.Codespace())
+	assert.Equal(t, types.CodeInvalidStatus, err.Code())
+
+	// upgrade!
+	codec.UpgradeFeatureMap[codec.AppTransferKey] = -1
+
+	// transfer: apps[0]-->apps[3](new) - success
+	err = transferApp(t, &ctx, &keeper, pubKeys[0], pubKeys[3])
+	assert.Nil(t, err)
+
+	// transfer: apps[3]-->apps[2](unstaking) - fail
+	err = transferApp(t, &ctx, &keeper, pubKeys[3], pubKeys[2])
+	assert.NotNil(t, err)
+	assert.Equal(t, keeper.Codespace(), err.Codespace())
+	assert.Equal(t, types.CodeInvalidStatus, err.Code())
+
+	// transfer: apps[3]-->apps[0](previous owner) success
```

Comment:
1. Why do we expect this to succeed?
2. If we were to wait for a new block, would this fail since the previous app is never unstaked?

Reply:
Can you update the test to verify that IF we do wait for a block, the transfer back fails?

Reply:
> Can you update the test to verify that IF we do wait for a block, the transfer back fails?

If a block was generated after the line 462, the test would still pass.  After the app ownership is transferred from A to B, the address A is the same as being unstaked or never being staked, meaning the address can be staked again.